### DISCUSSION TITLE
How to customize the Eclipse Theia Menu item.

### DISCUSSION THREAD

**Discussion by xinzhang20241216** [eclipse-theia/theia#14664] [General] [answered]
Whether Eclipse Theia support the Menu item customized, like remove the menu item file/Edit/Go
<img width="960" alt="image" src="https://github.com/user-attachments/assets/4ef7347c-a845-45a5-9e38-299eb74803c8" />

**Comment by msujew** [comment, marked as answer]
Hey @xinzhang20241216,

You can use the `MenuModelRegistry#unregister...` methods. See how to create menu contributions [in our documentation](https://theia-ide.org/docs/commands_keybindings/#contributing-menu-items).

**Reply by xinzhang20241216** [reply]
Hi @msujew, thanks for your suggestion, I also refer to below link and successfully remove the File Menu successfully.

https://github.com/eclipse-theia/theia/discussions/9577

The code is something like below, in one extension's menu contribution class-> registerMenus, just using `unregisterMenuAction(CommonMenus.FILE.slice(-1)[0]);` would be OK, but is there any best practice to unregister the menu item from global level instead of from some extension?

```
@injectable()
export class HelloWorldMenuContribution implements MenuContribution {

    registerMenus(menus: MenuModelRegistry): void {
        menus.unregisterMenuAction(CommonMenus.FILE.slice(-1)[0]);
    }
}
```

<img width="781" alt="Theia no file" src="https://github.com/user-attachments/assets/a2341589-8e81-4f5c-9c24-658ca085f4fa" />

**Reply by msujew** [reply]
> is there any best practice to unregister the menu item from global level instead of from some extension?

What do you mean with "global level"? Can you elaborate?

**Reply by xinzhang20241216** [reply]
OK, Assume that we have one Eclipse Theia application, the project structure is like below:
--browser-app
&nbsp;&nbsp;--src-gen
&nbsp;&nbsp;&nbsp;&nbsp; --backend
 &nbsp;&nbsp;&nbsp;&nbsp;--frontend
--electron-app
--some extension
&nbsp;&nbsp;--src
&nbsp;&nbsp;&nbsp;&nbsp;--backend
&nbsp;&nbsp;&nbsp;&nbsp;--frontend

My understanding is Eclipse Theia skeleton part(global level) is in the browser-app, it will dynamicly load the extension part. so my question is when we want to unregister some menu, is it possible that we do not implement it in one extension, but implement it in the browser-app skeleton part(global level)? or whether there are some listener that we can listen the page load or the page lifecycle, so in different stage of page lifecycle, we can do the menu unregister or register.

**Reply by msujew** [reply]
> it will dynamically load the extension part

That is incorrect, all Theia extensions are statically compiled into the final bundle.

> but implement it in the browser-app skeleton part(global level)?

You cannot implement anything in Theia outside of an extension. At least not by design, and everything else is a hack.

> or whether there are some listener that we can listen the page load or the page lifecycle, so in different stage of page lifecycle, we can do the menu unregister or register.

Refer to the [FrontendApplicationContribution](https://eclipse-theia.github.io/theia/docs/next/interfaces/core.FrontendApplicationContribution-1.html) and its various lifecycle hooks.

**Reply by xinzhang20241216** [reply]
Great thanks! now clear to me.
